DM-39876: Add bandaid for pydantic v2

timj opened the pull request:
Pydantic v2 breaks everything. Rather than pinning to <2 instead use the facility provided by v2 to import v1 APIs. This will give us some flexibility on timeline for fixing things to work only with v2.

## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

timj committed:
Provide None defaults for pydantic parameters

timj committed:
Try pydantic v1 API explicitly

pydantic v2 is not compatible with daf_butler at the moment.
Change all the imports to first try to load the v1 API explicitly
before falling back to the default API. pydantic.v1 works with
v2.

andy-slac commented on python/lsst/daf/butler/core/logging.py:
```diff
@@ -175,10 +179,10 @@ class ButlerLogRecord(BaseModel):
     filename: str
     pathname: str
     lineno: int
-    funcName: str | None
+    funcName: str | None = None
     process: int
     processName: str
-    exc_info: str | None
+    exc_info: str | None = None
     MDC: dict[str, str]
```
Does this work with v1? I thought that pydantic (v1) does not allow non-default attributes after default ones (but I may be thinking of dataclasses)?

timj replied:
I made that change first with the auto updater and it seemed to be okay. I am running Jenkins over night on everything so I'll see whether everything works in the morning.

andy-slac reviewed (approved):
Looks good, one minor question.

timj committed:
Declare new pydantic pin

<3 to protect us against the v3 breakage.

timj commented:
I'm not sure what to do about the mypy failures. Mypy doesn't know about pydantic.v1 and with pydantic v1 it complains about thta type not existing because it never falls back to the except block. With pydantic v2 installed it gets even more confused.

timj commented:
In theory changing mypy.ini to use `pydantic.v1` with pydantic v2 and then adding `type: ignore` on the fallback imports does work. Unfortunately it doesn't work at the moment because fastapi pins pydantic to v1 so when we install that pydantic reverts to v1 and then mypy gets confused about the `pydantic.v1` import (which we have to do first), since mypy always picks up the first import it sees.

timj committed:
Remove pkg_resources from types install

timj committed:
Remove mention of pkg_resources in docs

timj committed:
Add type: ignore for fallback import

This will only work in pydantic v2 with mypy but typing with
pydantic is already broken with v1 with the changes on this
branch.
